### ISSUE TITLE
duplicate file entry in .tar causes ratermount to fail?

### ISSUE THREAD

**Issue by soichih** [mxmlnkn/ratarmount#47]
I come across a .tar file that contains duplicate file entries like this.

```
tar --list -f files.tar
mask.nii.gz
.brainlife.json
.brainlife.json
```

I can untar this and I believe the last file overwrites the 2nd `.brainlife.json.

```
$ tar -xf files.tar
$ ls -a
.  ..  .brainlife.json  files.tar  mask.nii.gz
```

However, when I mount this .tar file with ratar mount, something goes wrong with the duplicated file. 

```
$ ls -la
ls: cannot access '.brainlife.json': Invalid argument
total 0
dr-x------  2 brlife brlife 204800 Jul 21  2019 .
drwxr-xr-x 10 root   root        0 Jan  5 21:26 ..
??????????  ? ?      ?           ?            ? .brainlife.json
-r--r--r--  2 833400 817579 178716 Jul 21  2019 mask.nii.gz
$ cat .brainlife.json
cat: .brainlife.json: Invalid argument
```

I know that tar file shouldn't have duplicate files to begin with, would it be possible to make ratarmount so that it can somehow handle this (maybe ignore duplicate files, or make the last/first file accessible?)

**Comment by mxmlnkn** [comment]
That's weird. This is definitely allowed and I even have [tests](https://github.com/mxmlnkn/ratarmount/blob/master/tests/updated-file.tar) for this. Can you reproduce the problem with that example TAR? The last file version will be shown by default and all other versions can be accessed using the special hidden folder `<filename>.versions`, see [here](https://github.com/mxmlnkn/ratarmount#file-versions).

Which ratarmount version are you using?
Can you try to force the index recreation with `-c`?
Could you send me a minimal non-working example for such a tar somehow. I tried creating one with the same file names and couldn't reproduce it.

**Comment by soichih** [comment]
Oh that's awesome > multiple files already supported.

I was using 0.5.0, but I just upgraded to 0.7.0 (`pip3 install ratarmount --upgrade`). Unfortunately, I am still having the same problem. Here is the test .tar file that I am having this problem

> http://dev1.soichi.us/tmp/file.tar

Maybe this .tar file itself might have some issue? 

To recreate the problem, I did

1) pip3 install ratarmount
2) download the .tar file, and run `ratarmount file.tar test`
3) run `cat test/.brainlife.json and I see the following error message.

```
cat: test/.brainlife.json: Invalid argument

```

**Comment by mxmlnkn** [comment]
Using `tar  tvlf file.tar`, I get:

```
-rw-r--r-- 833400/817579 78468 2019-10-08 21:38 mask.nii.gz
-rw-rw-r-- 833400/817579  7753 2019-10-08 22:54 .brainlife.json
hrw-rw-r-- 833400/817579     0 2019-10-08 22:54 .brainlife.json link to .brainlife.json
```
So, the second .brainlife.json file is actually a link to itself, or presumable to the older version with the same name. I think this might be an invalid TAR.

**Comment by soichih** [comment]
Ah~ I see. That's really messed up. I didn't know about -v option to show the file type!

I can't think of a way to workaround this on our end, but I will sleep on it.

**Comment by mxmlnkn** [comment]
How do you even create such a tar? Is it a common usecase?

Actually there are multiple bugs in ratarmount triggered because of this.
After fixing those and adding a rudimentary line to redirect self-links to older versions, it would work. It might be better than giving an error but still. Very unusual usecase.

The develop branch contains a quick and dirty fix. You can try it with:

    pip install git+https://github.com/mxmlnkn/ratarmount.git@develop#egginfo=ratarmount

Hope it helps for now.

**Comment by soichih** [comment]
Thank for the quick fix. I don't know if I install it right (still showing 0.7.0 as version) but I am still seeing the problem. I will try reinstalling it again later.

I really don't know how the broken .tar is created, but I use npm's archiver module combined with rsync / un-tar / tar commands that are executed in series between various computing resources. It's very convoluted in our case.

**Comment by mxmlnkn** [comment]
Yeah, sry. I didn't increase the version yet. So, pip might think it unnecessary to update. Try with:

    pip install --force-reinstall git+https://github.com/mxmlnkn/ratarmount.git@develop#egginfo=ratarmount

**Comment by soichih** [comment]
OMG it works! You are amazing. Thanks a lot!

**Comment by mxmlnkn** [comment]
I noticed something interesting today when trying to add a test for this. I already have a [test tar](https://github.com/mxmlnkn/ratarmount/blob/master/tests/file-existing-as-non-link-and-link.tar) from two years ago in this repo but it is not used in the tests.

Fortunately, I even documented how I created that file in [`createTests.sh`](https://github.com/mxmlnkn/ratarmount/blob/master/tests/create-tests.sh). These calls do create such files:

```
echo foo > bar
tar cf self-link.tar --owner=user --group=group bar bar
```

Or less weird looking:

```
mkdir foo
echo foo > foo/foo
tar cf self-link.tar foo/ foo/foo
```

Basically, it seems like when adding files twice, tar recognizes duplicates but and hardlinks them. Imo, it should also have recognized that the paths are identical though and simply skip it the second time.

All this means, these files might be more common than they should be.

**Comment by mxmlnkn** [comment]
As feared, it is even possible to create a tar containing only the self-link, which would with the current ratarmount quickfix enter an infinite loop:

```
echo foo > bar
tar cf self-link.tar --owner=user --group=group bar bar
tar --delete --occurrence=1 --file self-link.tar bar
tar tvlf self-link.tar
    hrwx------ user/group        0 2021-06-27 17:19 bar link to bar
```
Mounting this file with archivemount and then trying to list the mountpoint, actually makes archivemount hang up in an infinite loop and taking 100% cpu usage on one core :smile:.

**Comment by mxmlnkn** [comment]
Closed with 05f1cebf67870825cad2af8d8531627f04ac9bbd.
